cmd/agent/daemon/enrichment/enrichers.go

```diff
@@ -63,37 +68,45 @@ func (enricher *fileHashEnricher) Enrich(ctx context.Context, req *EnrichedConta
 		return
 	}
 
+	metricLabels := metricLabelsFromContainer(req.EbpfEvent.Container)
+	enrichMetric := metrics.AgentEnricherEventsTotalEnriched.WithLabelValues(metricLabels...)
+	enrichProcMissingMetric := metrics.AgentFileHashEnricherProcMissingTotal.WithLabelValues(metricLabels[1:]...)
+	enrichErrorMetric := metrics.AgentEnricherEventsTotalErrors.WithLabelValues(metricLabels...)
+
 	sha, err := enricher.calcFileHashForPID(req.EbpfEvent.Container, proc.PID(req.EbpfEvent.Context.NodeHostPid), exec.Path)
-	if err != nil {
-		if errors.Is(err, ErrFileDoesNotExist) {
-			return
-		}
-	} else {
+	if err == nil {
 		setExecFileHash(exec, sha)
+		enrichMetric.Inc()
 		return
 	}
+	enrichProcMissingMetric.Inc()
 
+	// If we can't find the file hash from the event PID, we try to find it from other PIDs in the same mount namespace
 	for _, pid := range enricher.mountNamespacePIDStore.GetBucket(proc.NamespaceID(req.EbpfEvent.Context.MntID)) {
 		if pid == proc.PID(req.EbpfEvent.Context.NodeHostPid) {
 			// We already tried that PID of the event, skipping.
 			continue
 		}
 
 		sha, err := enricher.calcFileHashForPID(req.EbpfEvent.Container, pid, exec.Path)
-		// We search for the first PID we can successfully calculate a filehash for.
 		if err != nil {
-			if errors.Is(err, ErrFileDoesNotExist) {
```

Comment:
This check is still valid IMHO, as the first PID we iterate will be the init process and if it returns a `ErrFileDoesNotExist`, we won't find it in any of the other procs.

Reply:
Should we only check the first PID then, instead of iterating trough the entire list? Actually, instead of reading the exec file from the process PID the first time and fallback to namespace PIDs, should we read from the root PID directly?

If I'm not mistaken, if we have nested namespaces the root PID will be valid as long as the child namespace exists. If the child namespace is deleted I don't think if there is anything else we can do.

Reply:
The only reason I do not 100% like this is, that it kinda relies on implicit behavior, that the first PID will always be the root hmm. I guess we can do it and add a comment 😅

Reply:
After thinking what is the best approach I have decided to leave as it was before, with the `ErrFileDoesNotExist` check